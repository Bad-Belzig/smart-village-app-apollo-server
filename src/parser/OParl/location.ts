import { isObjectLike } from 'lodash';

import { parseDate } from '../dateParser';

// only the necessary validation for parsing is done here.
// the created database object will be validated after creation before saving
// by using the .validate method that checks against the provided schema
export const parseLocation = (json) => {
  if (isObjectLike(json)) {
    const externalId = json.id; // change the id given to us to be the externalId instead
    delete json.id; // remove the id key from the json to ensure we do not preset it for mongo db

    let geojson: string | undefined;

    try {
      geojson = JSON.stringify(json.geojson);
    } catch (e) {
      geojson = undefined;
      console.log(`Invalid geojson for ${externalId}`, e);
    }
    return {
      ...json,
      externalId,
      created: parseDate(json.created),
      modified: parseDate(json.modified),

      geojson,
    };
  }
};
