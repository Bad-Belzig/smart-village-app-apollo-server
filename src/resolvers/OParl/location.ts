import {
  Body,
  ILocation,
  Location,
  Meeting,
  Organization,
  Paper,
} from '../../models';

export const locationResolvers = {
  Query: {
    oParlLocations: () => Location.find(),
  },
  OParlLocation: {
    bodies: (args: ILocation) =>
      args.bodies?.map((value) => Body.findOne({ externalId: value })),
    organization: (args: ILocation) =>
      args.organization?.map((value) =>
        Organization.findOne({ externalId: value }),
      ),
    meeting: (args: ILocation) =>
      args.meeting?.map((value) => Meeting.findOne({ externalId: value })),
    papers: (args: ILocation) =>
      args.papers?.map((value) => Paper.findOne({ externalId: value })),
  },
};
