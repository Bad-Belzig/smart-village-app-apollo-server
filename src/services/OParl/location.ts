import { fetch } from 'apollo-env';

import { Location } from '../../models';
import { parseLocation } from '../../parser';

// import flow
// 1. fetch json
// 2. create db object
//   2.1. validation necessary for parsing (done in parser)
//   2.2. parse json
//   2.3. create object
// 3. validate db object
// 4. queue new related objects
// 5. save to db

export const importLocation = async (url: string) => {
  try {
    const response = await fetch(url);

    if (response.ok) {
      const json = await response.json();

      // TODO: check for previous existence

      const lt = new Location(parseLocation(json));

      await lt.validate();

      // TODO: check for bodies, organizations, meetings and papers to import

      return lt.save();
    } else {
      throw new Error(
        `Error while fetching Location from : ${url}! \n response.status: ${response.status}`,
      );
    }
  } catch (e) {
    console.log(e);
  }
};
