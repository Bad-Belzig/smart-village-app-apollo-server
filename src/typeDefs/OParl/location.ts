import { gql } from 'apollo-server-core';

import { oParlBase } from './oParlBase';

export const locationGQL = gql`
  extend type Query {
    oParlLocations(externalIds: [String!]): [OParlLocation!]
  }

  type OParlLocation {
    bodies(offset: Int, pageSize: Int): [OParlBody!]
    description: String
    geojson: String
    locality: String
    meeting(offset: Int, pageSize: Int): [OParlMeeting!]
    organization(offset: Int, pageSize: Int): [OParlOrganization!]
    papers(offset: Int, pageSize: Int): [OParlPaper!]
    persons(offset: Int, pageSize: Int): [OParlPerson!]
    postalCode: String
    room: String
    streetAddress: String
    subLocality: String
    ${oParlBase}
  }
`;
