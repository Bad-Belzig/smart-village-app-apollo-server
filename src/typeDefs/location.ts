import { gql } from 'apollo-server-core';

export const locationGQL = gql`
  type Location {
    lat: String
    lon: String
  }
`;
