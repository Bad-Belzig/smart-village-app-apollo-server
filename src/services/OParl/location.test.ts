import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../testHelpers';
import { importLocation } from './location';

const helperFuncs = setupTestDataBase('test_location_import');

// TODO: change test url to something within our control
const testUrl =
  'https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/location/1-7';

describe('importing a location term', () => {
  beforeAll(helperFuncs.beforeAll);

  it('imports properly', basicImportTest(importLocation, testUrl));

  it('import returns undefined on faulty url', invalidUrlTest(importLocation));

  afterAll(helperFuncs.afterAll);
});
