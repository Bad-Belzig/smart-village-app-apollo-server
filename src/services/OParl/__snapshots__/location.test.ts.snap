// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a location term imports properly 1`] = `
Object {
  "__v": 0,
  "bodies": Array [
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1",
  ],
  "created": 2017-01-30T17:23:46.000Z,
  "description": "An Haus Clee 32, 41366 Schwalmtal",
  "externalId": "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/location/1-7",
  "keyword": Array [],
  "locality": "Schwalmtal",
  "meeting": Array [],
  "modified": 2020-12-11T10:51:00.000Z,
  "organization": Array [],
  "papers": Array [],
  "postalCode": "41366",
  "streetAddress": "An Haus Clee 32",
  "type": "https://schema.oparl.org/1.0/Location",
}
`;
